tfhe/src/high_level_api/tag.rs

```diff
@@ -0,0 +1,420 @@
+        formatter.write_str("a slice of bytes (&[u8]) or Vec<u8>")
+    }
+
+    fn visit_bytes<E>(self, bytes: &[u8]) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        let mut vec = SmallVec::default();
+        vec.set_data(bytes);
+        Ok(vec)
+    }
+
+    fn visit_byte_buf<E>(self, bytes: Vec<u8>) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(SmallVec::from_vec_conservative(bytes))
+    }
+}
+
+impl Versionize for SmallVec {
+    type Versioned<'vers> = &'vers [u8] where Self: 'vers;
+
+    fn versionize(&self) -> Self::Versioned<'_> {
+        self.data()
+    }
+}
+
+impl VersionizeOwned for SmallVec {
+    type VersionedOwned = Vec<u8>;
+
+    fn versionize_owned(self) -> Self::VersionedOwned {
+        match self {
+            Self::Stack { bytes, len } => bytes[..usize::from(len)].to_vec(),
+            Self::Heap(vec) => vec,
+        }
+    }
+}
+
+impl Unversionize for SmallVec {
+    fn unversionize(versioned: Self::VersionedOwned) -> Result<Self, UnversionizeError> {
+        Ok(Self::from_vec_conservative(versioned))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SmallVec {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_bytes(SmallVecVisitor)
+    }
+}
+
+/// Tag
+///
+/// The `Tag` allows to store bytes alongside of entities (keys, and ciphertext)
+/// the main purpose of this system is to `tag` / identify ciphertext with their keys.
+///
+/// tfhe-rs does not interpret or check this data, it only stores it and passes it around
+/// like so:
+///
+/// * When encrypted, a ciphertext gets the tag of the key used to encrypt it.
+/// * Ciphertexts resulting from operations (add, sub, etc.) get the tag from the ServerKey used
+/// * PublicKey gets its tag from the ClientKey that was used to create it
+/// * ServerKey gets its tag from the ClientKey that was used to create it
+///
+/// User can change the tag of any entities at any point.
+///
+/// # Example
+///
+/// ```
+/// use rand::random;
+/// use tfhe::prelude::*;
+/// use tfhe::{ClientKey, ConfigBuilder, FheUint32, ServerKey};
+///
+/// // Generate the client key then set its tag
+/// let mut cks = ClientKey::generate(ConfigBuilder::default());
+/// let tag_value = random();
+/// cks.tag_mut().set_u64(tag_value);
+/// assert_eq!(cks.tag().as_u64(), tag_value);
+///
+/// // The server key inherits the client key tag
+/// let sks = ServerKey::new(&cks);
+/// assert_eq!(sks.tag(), cks.tag());
+///
+/// // Encrypted data inherits the tag of the encryption key
+/// let a = FheUint32::encrypt(32832u32, &cks);
+/// assert_eq!(a.tag(), cks.tag());
+/// ```
+#[derive(
+    Default, Clone, Debug, serde::Serialize, serde::Deserialize, Versionize, PartialEq, Eq,
+)]
+#[versionize(TagVersions)]
+pub struct Tag {
+    // We don't want the enum to be public
+    inner: SmallVec,
+}
```

Comment:
this has a cost even if the Tag is empty given the default SmallVec is an array of bytes

Reply:
It has the  ~same cost as if we just had a Vec<T>, as the enum maybe like be just one byte more than an empty Vec<T>

Reply:
right